# Pengolahan Data Terstruktur dengan Pandas {#sec-modul-2}

::: {.callout-important title="Sub-CPMK Pertemuan Ini"}
Mahasiswa mampu mendemonstrasikan pengolahan data terstruktur dasar menggunakan Pandas, mengidentifikasi unit amatan dan unit analisis dari dataset perkotaan, serta memecahkan masalah kebingungan urutan eksekusi sel (**state** bersama) pada Jupyter Notebook.
:::

## Pendahuluan

[\<masukkan narasi yang menceritakan 'big why' mengelola data perkotaan yang terstruktur di sini\>]{.teks-placeholder style="color: #8a8a8a;"}

## Gambaran Umum Kasus

Pemerintah Kota Bandar Lampung hendak mengalokasikan bantuan pembangunan, dan pertanyaan pertamanya sederhana: kelurahan mana yang paling membutuhkan? Untuk menjawabnya, kita sebagai perencana memerlukan profil sosio-demografi setiap kelurahan — berapa kepadatan penduduknya, seberapa lengkap fasilitas pendidikan usia dininya, dan seterusnya. Bahan bakunya bersumber dari data **PODES 2021** (BPS) Kota Bandar Lampung. Data tersebut disimpan dalam format CSV.

Di studio, data seperti ini biasa kita rekap secara mandiri ke dalam tabel Excel saat melakukan kompilasi data. Kali ini kita berkenalan dengan **Pandas**, *library* Python untuk mengolah tabel rekapitulasi semacam itu. Namun sebelum menyentuh data, kita tuntaskan dulu janji Modul 1: **membuktikan cara kerja sel Jupyter Notebook** yang perilakunya berlawanan dengan dua berkas `.py` kita. Setelah jembatan itu terlewati, kita akan membaca rekap data CSV ke dalam Pandas, mengenali anatomi tabelnya, membedakan kelurahan sebagai *unit amatan* dari kecamatan sebagai *unit analisis*, merapikan tabel dengan membuat-menyusun-menghapus kolom, dan menutup dengan menyaring kelurahan berdasarkan kriteria untuk menemukan kandidat prioritas bantuan.


## Instruksi

Buat notebook baru `modul/modul-02.ipynb` di direktori proyek Anda, sambungkan ke *kernel* `analitika-perkotaan`, dan kerjakan seluruh instruksi di dalamnya.

### Jembatan Jupyter: Membuktikan Perilaku Sel dan **State** Bersama

Di Modul 1, `hitung_pincang.py` gagal karena setiap eksekusi `python berkas.py` adalah sesi kerja baru yang tidak mengingat apa pun. Sekarang kita reproduksi kasus yang sama di notebook — tetapi kali ini "berkas pincang"-nya kita pecah menjadi **dua sel**. Jalankan sel pertama:

In [1]:
luas_wilayah_km2 = 197.22
jumlah_penduduk = 1166066

Lalu jalankan sel kedua, yang hanya berisi perhitungan — persis si `hitung_pincang.py`:

In [2]:
kepadatan = jumlah_penduduk / luas_wilayah_km2
print(f"Kepadatan penduduk: {kepadatan:.0f} jiwa per km persegi")

Kepadatan penduduk: 5913 jiwa per km persegi


**Berhasil!** Padahal sel kedua tidak mendefinisikan variabelnya sendiri. Inilah perbedaan mendasarnya: sel-sel notebook **bukan** berkas-berkas terpisah, melainkan potongan-potongan percakapan dalam **satu sesi yang sama** — seperti saat kita mengetik berurutan di mode interaktif `>>>` kemarin, hanya saja kini tiap potongannya tersimpan rapi.

Analogi studionya: **kernel** adalah **papan tulis studio** tempat kita menulis bersama. Di sel pertama, kita menuliskan dua angka di papan; di sel kedua kita tinggal menoleh ke papan yang sama untuk memakainya tanpa harus menulis ulang. Segala coretan yang telah kita tulis di papan — variabel yang pernah kita isi — akan tetap bertahan di sana dan siap kita gunakan di sel mana pun. Kondisi papan tulis pada satu waktu tertentu itulah yang disebut dengan ****state**** yang **dibagi bersama** oleh semua sel.


### Eksperimen Papan Tulis: Mengapa Urutan Eksekusi Sel Sangat Menentukan Hasil

Papan tulis bersama itu nyaman, tetapi menyimpan satu jebakan yang membingungkan hampir semua pemula. Mari kita masuk ke jebakannya dengan sadar. Jalankan sel berikut, yang mengganti isi variabel luas dengan angka yang jelas-jelas salah:

In [3]:
luas_wilayah_km2 = 20

Sekarang **kembalilah ke sel perhitungan kepadatan di atas dan jalankan ulang sel itu** (klik selnya, lalu `Shift+Enter`). Hasilnya melonjak menjadi lebih dari 58 ribu jiwa per km persegi — karena perhitungan menoleh ke papan tulis dan menemukan angka `20` yang baru. Di modul ini kita tuliskan ulang perhitungannya sebagai sel berikut agar lonjakannya terlihat:

In [4]:
kepadatan = jumlah_penduduk / luas_wilayah_km2
print(f"Kepadatan penduduk: {kepadatan:.0f} jiwa per km persegi")

Kepadatan penduduk: 58303 jiwa per km persegi


Perhatikan dua hal:

1. Hasil sebuah sel **tidak ditentukan oleh posisinya di halaman**, melainkan oleh **urutan eksekusinya**. Sel yang tampak lebih atas bisa saja dijalankan belakangan dan memakai nilai yang lebih baru.

2. Angka dalam tanda kurung siku di kiri sel (misalnya `[5]`) adalah **nomor antrean eksekusi**. Kalau nomornya melompat-lompat tidak berurutan dari atas ke bawah, hasil yang tampil bisa jadi berasal dari papan tulis yang keadaannya sudah berubah.

::: {.callout-warning title="Penting!"}
### Papan Tulis Tidak Menghapus Dirinya Sendiri {.unnumbered .unlisted}
Coretan di papan tulis studio akan bertahan sampai kita sendiri menghapusnya — meski kita sudah selesai bekerja. Begitu pula variabel di **kernel**: sekali terisi, ia bertahan dan memengaruhi sel mana pun yang dijalankan setelahnya, **meskipun sel yang mengisinya sudah kita hapus dari notebook**.

Karena itu, biasakan ritual studio ini: sebelum kita mengumpulkan atau mempercayai hasil akhir notebook kita, jalankan **Restart *kernel* & Run All** (hapus papan tulis, lalu kerjakan seluruh sel berurutan dari atas). Jika ada sel yang gagal, berarti notebook Anda belum lengkap-mandiri secara berurutan — persis prinsip berkas `.py` di Modul 1, kini dalam versi notebook.
:::

Sebelum lanjut, kembalikan papan tulis ke angka yang benar:


In [5]:
luas_wilayah_km2 = 197.22

### Membuka Rekap Data: Membaca File CSV dengan Pandas

Sekarang kasus intinya. Data PODES kelurahan Bandar Lampung tersimpan sebagai berkas **CSV** (*comma-separated values*): berkas teks biasa yang tiap barisnya adalah satu catatan dan tiap kolomnya dipisahkan koma — bentuk paling sederhana dari **data terstruktur**, data yang tertata dalam baris dan kolom bermakna tetap. Baris pertamanya berisi nama-nama kolom. Kalau dibuka dengan Notepad, isinya kira-kira:

```
kode_prov,nama_prov,kode_kab,nama_kab,kode_kec,nama_kec,kode_des,nama_des,...
18,Lampung,71,Kota Bandar Lampung,10,Teluk Betung Barat,5,Bakung,...
```

Membaca teks seperti itu dengan mata tentu melelahkan. Di sinilah **Pandas** bekerja: ia memuat CSV menjadi tabel rekapitulasi yang rapi — persis lembar rekap kompilasi data studio di Excel — yang di Pandas disebut ***DataFrame***.

Berkas kita berisi 34 kolom, padahal pertemuan ini hanya butuh sebagian. Maka sebelum membaca, kita siapkan dua variabel penampung: `path_data` untuk alamat berkasnya (ingat, notebook kita berada di folder `modul`, sehingga *path* relatifnya naik satu tingkat dulu dengan `..`) dan `kolom_pilihan` untuk daftar kolom yang ingin dibaca. Dengan begitu, perintah bacanya sendiri tetap pendek:

In [6]:
import pandas as pd

path_data = "../dataset-ganjil-2026-2027/bdl_kelurahan.csv"
kolom_pilihan = ["nama_kec", "nama_des", "status_daerah",
                 "kepadatan_pddk", "jml_tk_negeri", "jml_tk_swasta",
                 "jml_ra_ba_negeri", "jml_ra_ba_swasta"]

df = pd.read_csv(path_data, usecols=kolom_pilihan)
df.head()

             nama_kec            nama_des  status_daerah  jml_tk_negeri  \
0  Teluk Betung Barat              Bakung              1              0   
1  Teluk Betung Barat             Kuripan              1              0   
2  Teluk Betung Barat  Negeri Olok Gading              1              0   
3  Teluk Betung Barat         Sukarame Ii              2              0   
4  Teluk Betung Barat          Batu Putuk              1              0   

   jml_tk_swasta  jml_ra_ba_negeri  jml_ra_ba_swasta  kepadatan_pddk  
0              2                 0                 0            6479  
1              0                 0                 0           16321  
2              1                 0                 0            7761  
3              3                 0                 0            1514  
4              1                 0                 0            2214  

Tiga baris pekerjaan terjadi di sana. `import pandas as pd` memanggil *library* Pandas dari rak **environment** kita dan memberinya nama panggilan pendek `pd` — cukup sekali di awal notebook. `pd.read_csv(...)` membaca berkas pada `path_data`, mengambil hanya kolom-kolom dalam `kolom_pilihan` (lewat `usecols=`), lalu hasilnya kita tampung di variabel `df`. Terakhir, `df.head()` mengintip **lima baris pertama** tabel untuk memastikan bacaan kita benar.

::: {.callout-note title="Konsep Python"}
### Tipe Data *list* {.unnumbered .unlisted}
`kolom_pilihan` adalah sebuah ***list***: tipe data yang menampung **banyak nilai sekaligus dalam satu kotak, berurutan**. Penulisannya memakai kurung siku, dengan koma sebagai pemisah: `["nama_kec", "nama_des", ...]`. Kalau variabel biasa adalah satu kotak berlabel, **list** adalah **rak berlabel yang isinya bernomor urut**.

Nomor urut itu disebut **indeks**, dan — ini kekhasan Python — **dihitung mulai dari 0**: `kolom_pilihan[0]` berisi `"nama_kec"`, `kolom_pilihan[1]` berisi `"nama_des"`, dan seterusnya. Panjang **list** bisa ditanyakan dengan `len(kolom_pilihan)`.

**list** akan sangat sering kita pakai untuk menampung daftar nama kolom seperti ini — daripada mengetik daftar panjang berulang-ulang di banyak perintah, kita tulis sekali di satu variabel lalu sebut labelnya.
:::


### Mengenal Struktur Tabel: Anatomi Baris, Kolom, dan Tipe Data *DataFrame*

Rekap data yang baru tiba di studio selalu kita periksa dulu kelengkapannya sebelum dipakai. Untuk *DataFrame*, pemeriksaan pertamanya adalah `info()`:

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126 entries, 0 to 125
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   nama_kec          126 non-null    object
 1   nama_des          126 non-null    object
 2   status_daerah     126 non-null    int64 
 3   jml_tk_negeri     126 non-null    int64 
 4   jml_tk_swasta     126 non-null    int64 
 5   jml_ra_ba_negeri  126 non-null    int64 
 6   jml_ra_ba_swasta  126 non-null    int64 
 7   kepadatan_pddk    126 non-null    int64 
dtypes: int64(6), object(2)
memory usage: 8.0+ KB


Laporan itu padat informasi. Bacalah dari atas:

- `126 entries` — tabel berisi **126 baris**, dan setiap baris adalah **satu kelurahan**. Di sinilah bedanya dengan tabel Excel bebas: dalam data terstruktur, satu baris selalu mewakili satu satuan pencatatan yang sama.

- Daftar kolom beserta `Non-Null Count` — berapa baris yang **terisi** di tiap kolom. Semuanya `126 non-null`, artinya tidak ada data kosong pada kolom-kolom pilihan kita (kemewahan yang tidak selalu terjadi; menangani data kosong adalah porsi Modul 4).

- `Dtype` — **tipe data** tiap kolom, saudara dari tipe data skalar di Modul 1: `int64` untuk bilangan bulat, `float64` untuk bilangan berkoma, dan `object` untuk teks/string seperti nama kecamatan.

Untuk kolom-kolom angka, kita bisa langsung meminta ringkasan statistiknya:

In [8]:
df.describe()

       status_daerah  jml_tk_negeri  jml_tk_swasta  jml_ra_ba_negeri  \
count     126.000000     126.000000     126.000000             126.0   
mean        1.015873       0.103175       2.595238               0.0   
std         0.125483       0.330560       2.075297               0.0   
min         1.000000       0.000000       0.000000               0.0   
25%         1.000000       0.000000       1.000000               0.0   
50%         1.000000       0.000000       2.000000               0.0   
75%         1.000000       0.000000       4.000000               0.0   
max         2.000000       2.000000      10.000000               0.0   

       jml_ra_ba_swasta  kepadatan_pddk  
count        126.000000      126.000000  
mean           0.269841     9431.825397  
std            0.662271     8428.116609  
min            0.000000      627.000000  
25%            0.000000     3466.750000  
50%            0.000000     6450.500000  
75%            0.000000    12693.500000  
max            

`describe()` menghitung sekaligus: jumlah data (`count`), rata-rata (`mean`), simpangan baku (`std`), nilai terkecil-terbesar (`min`, `max`), dan kuartil (`25%`, `50%`, `75%`). Dari satu tabel ini saja sudah terbaca, misalnya, betapa timpangnya kepadatan penduduk antar-kelurahan: ada kelurahan yang hanya ratusan jiwa per km², ada yang lebih dari 40 ribu. Pendalaman statistik deskriptif semacam ini adalah menu utama Modul 4 — hari ini cukup berkenalan.

### Kelurahan vs Kecamatan: Membedakan Unit Amatan dan Unit Analisis

Perhatikan lagi: setiap baris tabel kita adalah satu **kelurahan** — pada level itulah BPS mencatat datanya. Dalam bahasa metodologi, kelurahan adalah **unit amatan** (*unit of observation*) dataset ini: satuan tempat data dicatat, seperti satu lembar kuesioner terisi untuk satu kelurahan.

Namun pertanyaan perencanaan tidak selalu dijawab pada level pencatatan. Ketika pimpinan bertanya "kecamatan mana yang paling minim fasilitas PAUD?", kesimpulannya diambil pada level **kecamatan** — itulah **unit analisis** (*unit of analysis*): satuan tempat kesimpulan dibuat, seperti lembar rekapitulasi per wilayah yang merangkum tumpukan kuesioner. Unit amatan dan unit analisis boleh sama, boleh berbeda; yang fatal adalah tidak sadar sedang bekerja di level yang mana.

Mari kenali kedua level itu di data kita. Satu kolom di *DataFrame* dipanggil dengan `df["nama_kolom"]`, dan `nunique()` menghitung banyaknya **nilai unik** (nilai yang berbeda-beda) di dalamnya:

In [9]:
print(f"Jumlah kelurahan (unit amatan) : {df['nama_des'].nunique()}")
print(f"Jumlah kecamatan               : {df['nama_kec'].nunique()}")

Jumlah kelurahan (unit amatan) : 126
Jumlah kecamatan               : 20


126 kelurahan tersebar di 20 kecamatan. Lalu berapa kelurahan yang dinaungi masing-masing kecamatan? `value_counts()` menghitung **frekuensi kemunculan** tiap nilai pada sebuah kolom — alat andalan untuk meringkas **data kategorikal** (data berupa golongan/kategori, bukan angka hitungan):

In [10]:
df["nama_kec"].value_counts()

nama_kec
Kemiling                9
Panjang                 8
Tanjung Karang Barat    7
Sukabumi                7
Kedamaian               7
Rajabasa                7
Tanjung Karang Pusat    7
Kedaton                 7
Sukarame                6
Labuhan Ratu            6
Way Halim               6
Telukbetung Timur       6
Enggal                  6
Teluk Betung Utara      6
Teluk Betung Selatan    6
Langkapura              5
Tanjung Senang          5
Tanjung Karang Timur    5
Bumi Waras              5
Teluk Betung Barat      5
Name: count, dtype: int64

Tabel kecil itu sudah merupakan rekapitulasi pada unit analisis kecamatan: Teluk Betung Selatan menaungi paling banyak kelurahan, dan seterusnya. Simpan baik-baik `value_counts()` di ingatan — di Latihan Mandiri ia akan dipakai untuk meringkas kategori yang lebih bermakna sosial-ekonomi, dan di Modul 3 kita akan membuat rekapitulasi per kecamatan yang lebih kaya menggunakan SQL.

### Merapikan Tabel: Membuat, Menyusun Ulang, dan Menghapus Kolom

Fasilitas pendidikan usia dini di data kita terpecah menjadi empat kolom: TK negeri, TK swasta, RA/BA negeri, dan RA/BA swasta (RA/BA adalah *raudatul athfal*/*bustanul athfal*, PAUD di bawah Kemenag). Untuk pertanyaan alokasi bantuan, yang lebih berguna adalah **totalnya**. Di lembar rekap studio, kita akan menambahkan satu kolom jumlahan di ujung tabel — di Pandas pun sama: menugaskan nilai ke nama kolom yang belum ada akan **membuat kolom baru**:

In [11]:
df["jml_paud"] = df["jml_tk_negeri"] + df["jml_tk_swasta"] + df["jml_ra_ba_negeri"] + df["jml_ra_ba_swasta"]
df.head(3)

             nama_kec            nama_des  status_daerah  jml_tk_negeri  \
0  Teluk Betung Barat              Bakung              1              0   
1  Teluk Betung Barat             Kuripan              1              0   
2  Teluk Betung Barat  Negeri Olok Gading              1              0   

   jml_tk_swasta  jml_ra_ba_negeri  jml_ra_ba_swasta  kepadatan_pddk  jml_paud  
0              2                 0                 0            6479         2  
1              0                 0                 0           16321         0  
2              1                 0                 0            7761         1  

Perhatikan dua hal: penjumlahan antar-kolom bekerja **serentak untuk seluruh 126 baris** tanpa kita menghitung satu-satu, dan kolom baru `jml_paud` muncul di **ujung kanan** tabel. Posisi di ujung kanan itu kurang enak dibaca — kolom sepenting itu layak berdiri di dekat identitas kelurahan. Menyusun ulang posisi kolom caranya sederhana: sodorkan ***list*** berisi urutan kolom yang kita mau (pemakaian ulang konsep *list* barusan):

In [12]:
urutan_kolom = ["nama_kec", "nama_des", "status_daerah",
                "kepadatan_pddk", "jml_paud",
                "jml_tk_negeri", "jml_tk_swasta",
                "jml_ra_ba_negeri", "jml_ra_ba_swasta"]

df = df[urutan_kolom]
df.head(3)

             nama_kec            nama_des  status_daerah  kepadatan_pddk  \
0  Teluk Betung Barat              Bakung              1            6479   
1  Teluk Betung Barat             Kuripan              1           16321   
2  Teluk Betung Barat  Negeri Olok Gading              1            7761   

   jml_paud  jml_tk_negeri  jml_tk_swasta  jml_ra_ba_negeri  jml_ra_ba_swasta  
0         2              0              2                 0                 0  
1         0              0              0                 0                 0  
2         1              0              1                 0                 0  

Terakhir, setelah totalnya ada, empat kolom pecahannya tidak kita perlukan lagi hari ini. Membiarkan kolom tak terpakai membuat tabel lebar dan mata cepat lelah — seperti membiarkan kertas kalkir bekas menumpuk di meja gambar. Kita buang dengan `drop()`, lagi-lagi lewat variabel *list* agar rapi:

In [13]:
kolom_dibuang = ["jml_tk_negeri", "jml_tk_swasta", "jml_ra_ba_negeri", "jml_ra_ba_swasta"]

df = df.drop(columns=kolom_dibuang)
df.head(3)

             nama_kec            nama_des  status_daerah  kepadatan_pddk  \
0  Teluk Betung Barat              Bakung              1            6479   
1  Teluk Betung Barat             Kuripan              1           16321   
2  Teluk Betung Barat  Negeri Olok Gading              1            7761   

   jml_paud  
0         2  
1         0  
2         1  

Terakhir, setelah totalnya ada, empat kolom pecahannya tidak kita perlukan lagi hari ini. Membiarkan kolom tak terpakai membuat tabel lebar dan mata cepat lelah — seperti membiarkan kertas kalkir bekas menumpuk di ruang kerja kita. Kita buang dengan `drop()`, lagi-lagi lewat variabel **list** agar rapi:


### Menyaring Baris: Filtrasi Data Kelurahan Berdasarkan Kriteria Tertentu

Tabel sudah rapi; kini saatnya mengajukan pertanyaan sang pemberi kasus: **kelurahan mana yang paling membutuhkan bantuan fasilitas PAUD?** Tafsiran pertama yang paling sederhana: kelurahan yang fasilitas PAUD-nya sangat sedikit — katakanlah satu atau tidak ada sama sekali.

Menyaring baris di Pandas polanya `df[kondisi]`: "dari tabel `df`, ambil baris-baris yang memenuhi kondisi". Kondisinya kita tulis sebagai perbandingan atas sebuah kolom:

In [14]:
kelurahan_minim_paud = df[df["jml_paud"] <= 1]
print(f"Jumlah kelurahan dengan PAUD <= 1: {len(kelurahan_minim_paud)}")
kelurahan_minim_paud.head()

Jumlah kelurahan dengan PAUD <= 1: 37


             nama_kec            nama_des  status_daerah  kepadatan_pddk  \
1  Teluk Betung Barat             Kuripan              1           16321   
2  Teluk Betung Barat  Negeri Olok Gading              1            7761   
4  Teluk Betung Barat          Batu Putuk              1            2214   
8   Telukbetung Timur             Perwata              1           21630   
9   Telukbetung Timur          Way Tataan              1             687   

   jml_paud  
1         0  
2         1  
4         1  
8         0  
9         0  

::: {.callout-note title="Konsep Python"}
### Operator Perbandingan {.unnumbered .unlisted}
`df["jml_paud"] <= 1` memakai **operator perbandingan** — tanda yang membandingkan dua nilai dan menjawab benar (`True`) atau salah (`False`):

| Operator | Arti | Contoh |
|---|---|---|
| `==` | sama dengan | `status == 1` |
| `!=` | tidak sama dengan | `nama != "Bakung"` |
| `>` | lebih besar | `kepadatan > 10000` |
| `<` | lebih kecil | `luas < 5` |
| `>` | lebih besar atau sama | `jml_paud >= 2` |
| `<=` | lebih kecil atau sama | `jml_paud <= 1` |

Perhatikan: **sama dengan ditulis dua kali** (`==`), karena tanda `=` tunggal sudah dipakai untuk penugasan variabel (Modul 1). Menukar keduanya adalah salah ketik yang sangat umum.

Saat dipakai pada kolom **DataFrame**, perbandingan diuji **serentak ke seluruh baris**, menghasilkan deretan `True`/`False` sepanjang tabel — dan `df[...]` menahan hanya baris-baris yang `True`. Itulah mekanika di balik penyaringan barusan.
:::


In [15]:
kriteria_padat = df["kepadatan_pddk"] > 10000
kriteria_minim = df["jml_paud"] <= 1

kelurahan_prioritas = df[kriteria_padat & kriteria_minim]
print(f"Jumlah kelurahan prioritas: {len(kelurahan_prioritas)}")
kelurahan_prioritas

Jumlah kelurahan prioritas: 18


                 nama_kec          nama_des  status_daerah  kepadatan_pddk  \
1      Teluk Betung Barat           Kuripan              1           16321   
8       Telukbetung Timur           Perwata              1           21630   
10      Telukbetung Timur  Kota Karang Raya              1           23796   
11   Teluk Betung Selatan     Gedong Pakuon              1           13067   
12   Teluk Betung Selatan            Talang              1           19524   
14   Teluk Betung Selatan      Teluk Betung              1           25011   
31   Tanjung Karang Timur     Tanjung Agung              1           37717   
33   Tanjung Karang Timur        Sawah Lama              1           35775   
34   Tanjung Karang Timur      Sawah Brebes              1           26055   
51   Tanjung Karang Pusat           Kaliawi              1           25700   
54   Tanjung Karang Pusat   Kaliawi Persada              1           13159   
55                 Enggal       Gunung Sari              1      

::: {.callout-note title="Konsep Python"}
### Operator Boolean `&` dan `|` {.unnumbered .unlisted}
Untuk menggabungkan dua kondisi, Python menyediakan **operator boolean** (dinamai dari aljabar *Boolean*, aljabar benar-salah):

- `&` berarti **dan** — hasilnya `True` hanya jika **kedua** kondisi benar.

- `|` berarti **atau** — hasilnya `True` jika **salah satu saja** kondisi benar.

Dalam Python murni kata kuncinya `and` dan `or`, tetapi **khusus untuk kolom Pandas kita wajib memakai lambang `&` dan `|`**, karena pengujian dilakukan serentak ke ratusan baris sekaligus. Satu aturan lagi: jika kondisinya ditulis langsung tanpa variabel penampung, tiap kondisi **harus dikurung**, misalnya `df[(df["kepadatan_pddk"] > 10000) & (df["jml_paud"] <= 1)]` — tanpa kurung, Python salah membaca urutannya. Menampung kondisi ke variabel seperti yang kita lakukan membuat kurung tak diperlukan dan kodenya jauh lebih terbaca.
:::


## Latihan Mandiri

Kerjakan di notebook baru `modul/latihan-02.ipynb` dengan *kernel* `analitika-perkotaan`. Alurnya sama persis dengan instruksi di atas; yang berganti hanya kolom yang dipakai — kali ini **pendidikan dasar** (SD dan SMP), ditambah satu kolom kategorikal.

1. **Jembatan sel.** Di dua sel terpisah: sel pertama mendefinisikan dua variabel angka apa saja versi Anda sendiri, sel kedua menghitung sesuatu dari keduanya. Buktikan keduanya berbagi papan tulis, lalu jalankan **Restart *kernel* & Run All** dan pastikan semuanya tetap berjalan berurutan.

2. **Membaca data.** Baca `bdl_kelurahan.csv` memakai variabel `path_data` dan `kolom_pilihan` yang berisi: `nama_kec`, `nama_des`, `status_daerah`, `kepadatan_pddk`, `lap_usaha_sebagian_warga`, `jml_sd_negeri`, `jml_sd_swasta`, `jml_smp_negeri`, `jml_smp_swasta`.

3. **Anatomi tabel.** Periksa dengan `info()` dan `describe()`. Sebutkan (di sel teks): berapa barisnya, kolom mana saja yang bertipe angka, dan adakah kolom yang datanya kosong.

4. **Unit amatan vs unit analisis.** Hitung banyaknya nilai unik kelurahan dan kecamatan, lalu buat rekapitulasi frekuensi **data kategorikal** `lap_usaha_sebagian_warga` (lapangan usaha sebagian besar warga) dengan `value_counts()`. Nilainya berupa kode kategori BPS — arti tiap kode dapat dilihat di kamus data PODES 2021 yang dibagikan dosen. Tuliskan satu kalimat: lapangan usaha kode apa yang paling banyak menjadi tumpuan warga kelurahan-kelurahan Bandar Lampung, dan pada unit apa (amatan atau analisis) kesimpulan kalimat Anda itu berdiri?

5. **Merapikan tabel.** Buat kolom baru `jml_dikdas` (jumlah fasilitas pendidikan dasar) dari penjumlahan empat kolom SD/SMP, susun ulang kolom sehingga `jml_dikdas` berdiri setelah `kepadatan_pddk`, lalu hapus empat kolom pecahannya memakai variabel *list*.

6. **Menyaring baris.** Temukan **kelurahan prioritas fasilitas pendidikan dasar**: kepadatan penduduk di atas **15.000** jiwa/km² **dan** `jml_dikdas` paling banyak **2**. Berapa kelurahan yang lolos? Tutup notebook Anda dengan satu paragraf sel teks yang menafsirkan daftar tersebut sebagaimana laporan singkat kepada pimpinan.